Normalize function type during validation

tmiasko opened the pull request:
During inlining, the callee body is normalized and has types revealed,
but some of locals corresponding to the arguments might come from the
caller body which is not. As a result the caller body does not pass
validation without additional normalization.

Closes #78442.

rust-highfive commented:
r? @davidtwco

(rust_highfive has picked a reviewer for you, use r? to override)

lcnr commented on compiler/rustc_mir/src/transform/validate.rs:
```diff
@@ -357,7 +357,9 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
                 }
             }
             TerminatorKind::Call { func, args, destination, cleanup, .. } => {
+                let param_env = self.param_env.with_reveal_all_normalized(self.tcx);
```
shouldn't `self.param_env` just be reveal all from the start here?

davidtwco reviewed (changes requested):
Implementation LGTM; can you add a test for #78442?

tmiasko commented:
* Extended one of existing test cases to cover #78442.
* Removed misleading compiletest annotations which had no effect.
* Modified validator to always use reveal_all_normalized as param env.

davidtwco commented:
@bors r+

bors commented:
:pushpin: Commit 99be78d135e73197e04221c139a219ea6436e72a has been approved by `davidtwco`

<!-- @bors r=davidtwco 99be78d135e73197e04221c139a219ea6436e72a -->
<!-- homu: {"type":"Approved","sha":"99be78d135e73197e04221c139a219ea6436e72a","approver":"davidtwco"} -->
